models/demos/falcon7b/tt/falcon_mlp.py

```diff
@@ -47,11 +66,236 @@ def __init__(
+        self.prefill_seq_len = model_config["MLP_SEQ_LEN"]
+
+        layer_name = f"{base_url}.{layer_num}"
+        dense_h_to_4h_str = f"{layer_name}.mlp.dense_h_to_4h.weight"
+        dense_4h_to_h_str = f"{layer_name}.mlp.dense_4h_to_h.weight"
+
+        custom_output_shape_h_to_4h = (
+            (1, 1, self.padding_value, 4 * self.padding_value)
+            if self.prefill_seq_len in [1024, 2048] and model_config["OPTIMIZED_MODE"]
+            else None
+        )
+        custom_output_shape_4h_to_h = (
+            (1, 1, 4 * self.padding_value, self.padding_value)
+            if self.prefill_seq_len in [1024, 2048] and model_config["OPTIMIZED_MODE"]
+            else None
+        )
+
+        self.dense_h_to_4h_weights = get_weights_cached(
+            devices,
+            model_config,
+            tt_cache_path,
+            dense_h_to_4h_str,
+            weight_config_str="DENSE_H_TO_4H_MM_WEIGHTS",
+            weights_to_cache=(torch.transpose(state_dict[dense_h_to_4h_str], -2, -1) if state_dict else None),
+            weights_dict=weights_dict,
+            custom_output_shape=custom_output_shape_h_to_4h,
+        )
+        self.dense_4h_to_h_weights = get_weights_cached(
+            devices,
+            model_config,
+            tt_cache_path,
+            dense_4h_to_h_str,
+            weight_config_str="DENSE_4H_TO_H_MM_WEIGHTS",
+            weights_to_cache=(torch.transpose(state_dict[dense_4h_to_h_str], -2, -1) if state_dict else None),
+            weights_dict=weights_dict,
+            custom_output_shape=custom_output_shape_4h_to_h,
+        )
+        if "MLP_DECODE_PADDING_TENSORS" not in self.model_config and model_config["OPTIMIZED_MODE"]:
+            self._load_mlp_padded_tensors()
+
+    def _load_mlp_padded_tensors(self):
+        tt_paddings = []
+        for device_id in range(len(self.devices)):
+            tt_padding = torch.zeros((1, 1, 32, 64)).bfloat16().float()  # 4608 - 4544 = 64, batch=32
+            tt_padding = ttnn.from_torch(
+                tt_padding,
+                ttnn.bfloat16,
+                device=self.devices[device_id],
+                layout=ttnn.TILE_LAYOUT,
+                memory_config=ttnn.DRAM_MEMORY_CONFIG,
+            )
+            tt_paddings.append(tt_padding)
+
+        self.model_config["MLP_DECODE_PADDING_TENSORS"] = tt_paddings
+
     def forward(self, x: tt_lib.tensor.Tensor) -> tt_lib.tensor.Tensor:
         hidden_states = []
         for device_id in range(len(x)):
+            # pad inputs with padding tensor if not already padded
+            if (
+                x[device_id].shape[-1] < self.padding_value
+                and self.prefill_seq_len in [1024, 2048]
+                and self.model_config["OPTIMIZED_MODE"]
+            ):
+                x[device_id] = ttnn.concat(
+                    [x[device_id], self.model_config["MLP_DECODE_PADDING_TENSORS"][device_id]], dim=3
+                )
```

Comment:
Does this padding affect decode perf? Could we not instead unpad the weights and do the MLP as before for decode?

Reply:
Concat and unpad add 11us of overhead so I believe it doesn't affect decode perf. On the other hand, it helps us optimize prefill MLP matmuls with convenient weight dims (for sharded params) and that's why we pad MLP weights. Here are unit test perf numbers:
[ops_perf_results_2024_04_25_12_46_16.csv](https://github.com/tenstorrent/tt-metal/files/15110400/ops_perf_results_2024_04_25_12_46_16.csv)